internal/controller/config_controller.go

```diff
@@ -106,28 +160,95 @@ func (r *ConfigReconcile) reconcile() error {
-	// for the current reconciled config resource
-	clusterHasOwnerRef := false
-	for i, _ := range r.cephCluster.OwnerReferences {
-		ownerRef := &r.cephCluster.OwnerReferences[i]
-		if ownerRef.UID == r.config.UID {
-			clusterHasOwnerRef = true
-			break
+	return nil
+}
+
+func (r *ConfigReconcile) reconcileCephCsiClusterInfo() error {
+	csiConfigMap := corev1.ConfigMap{}
+	csiConfigMap.Name = utils.CsiConfigVolume.Name
+	csiConfigMap.Namespace = r.config.Namespace
+
+	log := r.log.WithValues("csiConfigMapName", csiConfigMap.Name)
+	log.Info("Reconciling Ceph CSI Cluster Info")
+
+	// Creting the desired record in advance to miimize the amount execution time
+	// the code run while serializing access to the configmap
+	record := composeCsiClusterInfoRecord(&r.config, &r.cephCluster)
+
+	// Using a lock to serialized the updating of the config map.
+	// Although the code will run perfetcly fine without the lock, there will be a higher
+	// chance to fail on the create/update operation because another concurrent reconcile loop
+	// updated the config map which will result in stale representation and an update failure.
+	// The locking stategy will sync all update to the shared config file and will prevent such
+	// potential issues without a big impact on preformace as a whole
+	configMapUpdateLock.Lock()
+	defer configMapUpdateLock.Unlock()
+
+	_, err := ctrlutil.CreateOrUpdate(r.ctx, r.Client, &csiConfigMap, func() error {
+		if err := ctrlutil.SetOwnerReference(&r.config, &csiConfigMap, r.Scheme); err != nil {
+			log.Error(err, "Failed setting an owner reference on Ceph CSI config map")
+			return err
 		}
-	}
-	if !clusterHasOwnerRef {
-		ctrlutil.SetOwnerReference(&r.config, &r.cephCluster, r.Scheme)
-		r.log.Info("Owner reference missing on ceph cluster resource, adding")
-		if err := r.Update(r.ctx, &r.cephCluster); err != nil {
-			r.log.Error(err, "Failed adding an owner reference to ceph cluster resource")
+
+		configsAsJson := csiConfigMap.Data["config.json"]
+		clusterInfoList := []*csiClusterInfoRecord{}
+
+		// parse the json serialized list into a go array
+		if configsAsJson != "" {
+			if err := json.Unmarshal([]byte(configsAsJson), &clusterInfoList); err != nil {
+				log.Error(err, "Failed to parse cluster info list under \"config.json\" key")
+				return err
+			}
+		}
+
+		// locate an existing entry for the same id if exists
+		index := slices.IndexFunc(clusterInfoList, func(record *csiClusterInfoRecord) bool {
+			return record.ClusterId == r.config.Name
+		})
+
+		// overwrite an existing entry or append a new one
+		if index > -1 {
+			clusterInfoList[index] = record
+		} else {
+			clusterInfoList = append(clusterInfoList, record)
+		}
+
+		// serialize the list ba
+		if bytes, err := json.Marshal(clusterInfoList); err == nil {
+			if csiConfigMap.Data == nil {
+				csiConfigMap.Data = map[string]string{}
+			}
+			csiConfigMap.Data["config.json"] = string(bytes)
```

Comment:
define a const for `config.json`?

Reply:
I didn't see a reason, but you mentioned it so I will add a const.

Reply:
done